File: modal/src/main/java/ir/drax/modal/ModalBuilder.java
package ir.drax.modal;

import android.animation.Animator;
import android.animation.AnimatorListenerAdapter;
import android.graphics.drawable.TransitionDrawable;
import android.os.Build;
import android.view.Gravity;
import android.view.View;
import android.view.ViewGroup;
import android.view.ViewTreeObserver;
import android.view.animation.CycleInterpolator;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.core.content.res.ResourcesCompat;

import java.util.Observable;
import java.util.Observer;

import ir.drax.modal.model.Direction;
import ir.drax.modal.model.MoButton;
import ir.drax.modal.model.ModalObj;

public class ModalBuilder extends RelativeLayout implements Observer, androidx.lifecycle.Observer<ModalObj> {
    private ViewGroup root;

    private ModalObj modalObj;
    private FrameLayout bg;

    ModalBuilder(ModalObj data) {
        super(data.getRoot().getContext());

        this.modalObj = data;
        this.root = modalObj.getRoot();

        addView(modalObj.getModal(), new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.WRAP_CONTENT));

        if (modalObj.isBlurEnabled())
            initBlurEffect();
    }

    public void show(){
        buildModal();
        modalObj.observeForever( this);
    }

    public void hide(){
        if (bg != null)
            closeModal(bg);
    }

    private void buildModal(){

        if (bg == null) {
            bg = new FrameLayout(getContext());
            bg.setTag(modalObj.getViewIdTag());

            setTag(modalObj.getDirection());
            setViewDirection();

            setVisibility(View.INVISIBLE);

            setHeader(bg);

            if (modalObj.getType().equals(Modal.Type.Progress))
                setProgress();

            setReaction();


            if (modalObj.getType().equals(Modal.Type.List))
                setList();

            bg.setBackgroundDrawable(getResources().getDrawable(R.drawable.modal_root_transition));

            bg.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    closeModal(bg);
                }
            });

            bg.addView(this);
        }

        root.addView(bg,new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,ViewGroup.LayoutParams.MATCH_PARENT));

        root.getViewTreeObserver().addOnGlobalLayoutListener(new ViewTreeObserver.OnGlobalLayoutListener() {
            @Override
            public void onGlobalLayout() {
                ViewTreeObserver obs = root.getViewTreeObserver();

                FrameLayout.LayoutParams fLayoutParams = new FrameLayout.LayoutParams(FrameLayout.LayoutParams.MATCH_PARENT,FrameLayout.LayoutParams.WRAP_CONTENT);
                if (((int)getTag())== Direction.FromBottom) {
                    fLayoutParams.gravity = Gravity.BOTTOM;
                    fLayoutParams.bottomMargin = getHeight() * -1;

                }else{
                    fLayoutParams.gravity = Gravity.TOP;
                    fLayoutParams.topMargin = getHeight() * -1;
                }
                setLayoutParams(fLayoutParams);
                animate()
                        .setStartDelay(250)
                        .setListener(new AnimatorListenerAdapter() {
                            @Override
                            public void onAnimationStart(Animator animation) {
                                setVisibility(View.VISIBLE);

                                if (Build.VERSION.SDK_INT > Build.VERSION_CODES.LOLLIPOP) {
                                    TransitionDrawable transitionDrawable = (TransitionDrawable) bg.getBackground();
                                    transitionDrawable.startTransition(250);
                                }else
                                    bg.setBackgroundColor(getResources().getColor(R.color.modal_bg_dark_transparent));

                                super.onAnimationStart(animation);
                            }

                            @Override
                            public void onAnimationEnd(Animator animation) {
                                if (modalObj.getListener() != null)modalObj.getListener().onShow();

                                if (!modalObj.getType().equals(Modal.Type.Custom))
                                    findViewById(R.id.ok)
                                            .animate()
                                            .translationY(
                                                    ((int)getTag())==Direction.FromBottom?
                                                            -16:0)
                                            .setDuration(400)
                                            .setInterpolator(new CycleInterpolator(0.1f))
                                            .start();


                                super.onAnimationEnd(animation);
                                blurEffect(true);
                            }
                        })
                        .translationY(
                                ((int)getTag())==Direction.FromBottom?
                                        -getHeight() : getHeight())
                        .setDuration(500)
                        .start();

                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                    obs.removeOnGlobalLayoutListener(this);
                } else {
                    obs.removeGlobalOnLayoutListener(this);
                }
            }
        });
    }

    private void setProgress() {
        if (modalObj.getType().equals(Modal.Type.Custom))return;

        ProgressBar progress = findViewById(R.id.progress);
        ((TextView)findViewById(R.id.percentage)).setText(modalObj.getProgress()+"%");

        progress.setIndeterminate(modalObj.getProgress()==0);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            progress.setProgress(modalObj.getProgress(),true);

        }else progress.setProgress(modalObj.getProgress());
    }

    private void setList() {
        if (modalObj.getType().equals(Modal.Type.Custom))return;

        TextView doneBtnView = findViewById(R.id.ok);
        ViewGroup listHolder =  findViewById(R.id.listItems);
        LinearLayout.LayoutParams itemLP = new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,ViewGroup.LayoutParams.WRAP_CONTENT);
        for (final MoButton listItem : modalObj.getList()) {
            TextView textView = new TextView(getContext());
            textView.setText(listItem.getText());
            //textView.setText(Html.fromHtml(""));
            textView.setTextColor(getResources().getColor(R.color.black_faded));
            textView.setTypeface(doneBtnView.getTypeface());
            textView.setCompoundDrawablesWithIntrinsicBounds(0,0,listItem.getIcon(),0);
            textView.setCompoundDrawablePadding(8);
            //textView.tint(activity.getResources().getColorStateList();
            //textView.getCompoundDrawables()[0].setColorFilter(new PorterDuffColorFilter(ContextCompat.getColor(textView.getContext(), R.color.grey), PorterDuff.Mode.SRC_IN));

            textView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (listItem.getClickListener()!=null)listItem.getClickListener().onClick(v);
                }
            });
            //textView.setBackgroundColor(activity.getResources().getColor(listHolder.getChildCount()%2==0?R.color.modal_list_item_odd:R.color.modal_list_item_even));
            //textView.setBackgroundResource(R.drawable.list_item_bg);
            textView.setPadding(32,16,32,16);

            listHolder.addView(textView,itemLP);

            View divider = new View(getContext());
            divider.setBackgroundColor(getResources().getColor(R.color.black_faded));
            listHolder.addView(divider,new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,5));
        }
    }

    private void setReaction() {
        if (modalObj.getType().equals(Modal.Type.Custom))return;

        TextView doneBtnView = findViewById(R.id.ok);
        if (modalObj.getReAction() != null) {
            doneBtnView.setText(modalObj.getReAction().getText());
            doneBtnView.setCompoundDrawablesWithIntrinsicBounds(modalObj.getReAction().getIcon(),0,0,0);
            doneBtnView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (modalObj.getReAction().getClickListener()!=null)
                        if(modalObj.getReAction().getClickListener().onClick(v))
                            closeModal(bg);
                }
            });
            //doneBtnView.getCompoundDrawables()[0].setColorFilter(new PorterDuffColorFilter(ContextCompat.getColor(doneBtnView.getContext(), R.color.grey), PorterDuff.Mode.SRC_IN));
        }
    }

    private void setHeader(final FrameLayout bg) {
        if (modalObj.getType().equals(Modal.Type.Custom))return;

        if (!modalObj.getMessage().getText().toString().isEmpty()) {
            final TextView summary = findViewById(R.id.text);
            summary.setText(modalObj.getMessage().getText());
            summary.setCompoundDrawablesWithIntrinsicBounds(modalObj.getMessage().getIcon(),0,0,0);
            summary.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (modalObj.getMessage().getClickListener()!=null)
                        if (modalObj.getMessage().getClickListener().onClick(v))
                            closeModal(bg);
                }
            });
        }


        if (!modalObj.getTitle().toString().isEmpty())
            ((TextView)findViewById(R.id.title)).setText(modalObj.getTitle());

        if (modalObj.getIcon() > 0)
            ((ImageView)findViewById(R.id.icon)).setImageResource(modalObj.getIcon());
        else
            findViewById(R.id.icon).setVisibility(View.GONE);

    }

    void setViewDirection(){
        if (((int)getTag())==Direction.FromBottom)
            setBackgroundDrawable(ResourcesCompat.getDrawable(getResources(),R.drawable.top_curved_header,null));
        else
            setBackgroundDrawable(ResourcesCompat.getDrawable(getResources(),R.drawable.bottom_curved_header,null));

    }

    void initBlurEffect() {
        /*blurView=new BlurView(activity);
        blurView.setLayoutParams(new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT));
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            blurView.setupWith(root)
                    //.windowBackground(root.getBackground())
                    .blurAlgorithm(new RenderScriptBlur(activity))
                    .blurRadius(2)
                    .setHasFixedTransformationMatrix(true);
        }

        root.addView(blurView);*/
    }

    void blurEffect(boolean set) {
        if (!modalObj.isBlurEnabled())return;
        /*if (set) {
            blurView.setBlurEnabled(true);

        } else
            blurView.setBlurEnabled(false);*/

    }

    boolean closeModal( final View bg) {
        if (modalObj.isLockVisibility())return false;

        animate()
                .translationY(
                        ((int)getTag())==Direction.FromBottom?
                                getHeight() : -getHeight())
                .setDuration(250)
                .setListener(new AnimatorListenerAdapter() {
                    @Override
                    public void onAnimationEnd(Animator animation) {
                        if (Build.VERSION.SDK_INT > Build.VERSION_CODES.LOLLIPOP) {
                            TransitionDrawable transitionDrawable = (TransitionDrawable) bg.getBackground();
                            transitionDrawable.reverseTransition(250);
                        }
                        modalObj.removeObserver(ModalBuilder.this);
                        root.removeView(bg);
                        blurEffect(false);

                        if (modalObj.getListener()!=null)
                            modalObj.getListener().onDismiss();

                        super.onAnimationEnd(animation);
                    }
                })
                .start();
        return true;
    }

    public ModalObj state(){
        return modalObj;
    }

    @Override
    public void onChanged(ModalObj modalObj) {
        switch (modalObj.getChangedIndex()){
            case 0: case 1: case 2:
                setHeader(bg);break;

            case 3:
                setReaction();break;

            case 4:
                setList();break;

            case 5:
                setProgress();break;

        }
    }

    @Override
    public void update(Observable observable, Object o) {

    }
}

File: modal/src/main/java/ir/drax/modal/ModalBuilder.kt
package ir.drax.modal

import android.content.Context
import android.graphics.drawable.TransitionDrawable
import android.os.Build
import android.util.AttributeSet
import android.view.*
import android.view.ViewTreeObserver.OnGlobalLayoutListener
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import android.widget.*
import androidx.core.content.res.ResourcesCompat
import androidx.core.view.drawToBitmap
import androidx.interpolator.view.animation.FastOutSlowInInterpolator
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import ir.drax.modal.model.ModalObj
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.lang.Exception
import java.util.*

class ModalBuilder @JvmOverloads constructor(var options:ModalObj, val root: ViewGroup, context: Context=root.context, attrs: AttributeSet?=null, defStyleAttr: Int = 0)
    :RelativeLayout(context,attrs,defStyleAttr)
        ,Observer<ModalObj> {

    private var slideAnimation: Animation? = null
    private var bg:ViewGroup
    private var layoutListener: OnGlobalLayoutListener?=null

    private val blurBg :ImageView by lazy {
        ImageView(root.context)
    }
    private var onAnimationFinishedQueue:MutableList<() -> Unit> = mutableListOf()

    private fun drawBlurBg(){
        val screenToCapture = root.drawToBitmap()

        GlobalScope.launch(Dispatchers.IO) {
            try {
                BlurBuilder(root.context)
                        .blur(screenToCapture).let {bitmap ->

                            withContext(Dispatchers.Main){
                                blurBg.setImageBitmap(bitmap)
                            }
                        }

            }catch (e:Exception){
                e.printStackTrace()
            }
        }
    }

    init {
        addView(options.modal, LayoutParams(LayoutParams.MATCH_PARENT,LayoutParams.WRAP_CONTENT))
        tag=options.direction
        visibility= View.INVISIBLE
        bg =FrameLayout(root.context)
        bg.tag=ModalObj.VIEW_TAG_ID
        setViewDirection()
        setHeader(bg)
        setProgress()
        setCallback()
        setList()
        bg.background=ResourcesCompat.getDrawable(resources,R.drawable.modal_root_transition,null)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            bg.elevation=100f
        }
        bg.setOnClickListener{ hide(bg) }
        bg.addView(blurBg,ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,ViewGroup.LayoutParams.MATCH_PARENT))
        bg.addView(this,ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,ViewGroup.LayoutParams.WRAP_CONTENT))


        // Display direction set
        with(FrameLayout.LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT)){
            gravity =
                    if (options.direction == Modal.Direction.Bottom)
                        Gravity.BOTTOM
                    else
                        Gravity.TOP

            layoutParams = this
        }
    }

    fun hide() = hide(bg,true)

    fun show():ModalBuilder{
        if (slideAnimation != null){
            onAnimationFinishedQueue.add {
                show()
            }

        }else {
            buildModal()
            options.update.observeForever(this)
        }
        return this
    }

    private fun buildModal(){
        if(bg.parent ==null) {
            root.addView(bg, LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT))
            object : OnGlobalLayoutListener {
                override fun onGlobalLayout() {

                    root.viewTreeObserver.removeOnGlobalLayoutListener(this)

                    if (options.type == Modal.Type.List)
                        findViewById<RecyclerView>(R.id.listScrollView).apply {
                            if (measuredHeight > (bg.height * .7)) {
                                layoutParams.height = (bg.height * .7).toInt()
                                requestLayout()
                            }
                        }


                    // On Animation start
                    if(options.blurEnabled.not())
                        (bg.background as TransitionDrawable).startTransition(options.animationDuration.toInt())
                    else
                        blurEffect(true)
                    // On Animation start


                    visibility = View.VISIBLE

                    animateIn(options.direction){

                        options.listener?.onShow()
                        if (options.type != Modal.Type.Custom) {
                            animateOkBtn()
                        }

                    }
                }
            }.also { layoutListener = it }
            root.viewTreeObserver?.addOnGlobalLayoutListener(layoutListener)
        }
    }

    private fun animateOkBtn() {
        findViewById<View>(R.id.ok)
                .animate()
                .translationY(
                        if (options.direction == Modal.Direction.Bottom)
                            -16f
                        else
                            0f
                )
                .setDuration(options.animationDuration)
                .start()
    }

    fun hide(header:View, forceClose:Boolean=false) : Boolean {
        return if (slideAnimation != null){
            onAnimationFinishedQueue.add {
                hide(header, forceClose)
            }
            false

        }else if (options.lockVisibility.not() || forceClose){

            root.viewTreeObserver.removeOnGlobalLayoutListener(layoutListener)

            if(options.blurEnabled.not())
                (header.background as TransitionDrawable).resetTransition()
            else
                blurEffect(false)

            animateOut(options.direction){

                release(header)
                options.listener.let {
                    it?.onDismiss()
                }
            }

            true

        }else
            false
    }

    private fun release(header: View) {
        options.update.removeObserver(this@ModalBuilder)
        root.removeView(header)
    }

    private inline fun animateOut(direction: Modal.Direction, crossinline onEnd:()->Unit = {}) {
        val animation =
                when(direction){
                    Modal.Direction.Bottom -> R.anim.slide_exit_bottom
                    Modal.Direction.Top -> R.anim.slide_exit_top
                }

        slideAnimation = AnimationUtils
                .loadAnimation(context, animation)

        startAnimation(
                slideAnimation?.apply {
                    duration = options.exitAnimationDuration

                    setAnimationListener(object :Animation.AnimationListener{
                        override fun onAnimationStart(p0: Animation?) {}

                        override fun onAnimationEnd(p0: Animation?) {
                            animationFinished(onEnd)
                        }

                        override fun onAnimationRepeat(p0: Animation?) {}

                    })
                }
        )
    }

    private inline fun animationFinished(crossinline onEnd: () -> Unit) {
        onEnd.invoke()
        slideAnimation = null
        onAnimationFinishedQueue
                .lastOrNull()
                ?.let {
                    onAnimationFinishedQueue.clear()
                    it.invoke()
                }
    }

    private inline fun animateIn(direction: Modal.Direction, crossinline onEnd:()->Unit = {}) {
        val animation =
                when(direction){
                    Modal.Direction.Top -> R.anim.slide_enter_top
                    Modal.Direction.Bottom -> R.anim.slide_enter_bottom
                }

        slideAnimation = AnimationUtils
                .loadAnimation(context, animation)

        startAnimation(
                slideAnimation?.apply {

                    duration = options.animationDuration
                    interpolator = FastOutSlowInInterpolator()

                    setAnimationListener(object :Animation.AnimationListener{
                        override fun onAnimationStart(p0: Animation?) {}

                        override fun onAnimationEnd(p0: Animation?) {
                            animationFinished(onEnd)
                        }

                        override fun onAnimationRepeat(p0: Animation?) {}

                    })
                }
        )
    }

    private fun setHeader(header:ViewGroup){
        if (options.type != Modal.Type.Custom){
            if (options.message.displayText.isEmpty().not()){
                val summary=findViewById<TextView>(R.id.text)
                summary.text=options.message.displayText
                summary.setCompoundDrawablesWithIntrinsicBounds(options.message.iconResourceId,0,0,0)
                summary.setOnClickListener {
                    if(options.message.clickListener(summary))
                        hide(header,true)
                }
            }

            findViewById<TextView>(R.id.title).text = options.title

            if (options.icon==0) findViewById<ImageView>(R.id.icon).visibility= View.GONE
            else findViewById<ImageView>(R.id.icon).setImageResource(options.icon)

            if (options.lockVisibility)
                findViewById<ImageView>(R.id.close).visibility= View.INVISIBLE

        }
    }

    private fun setCallback(){
        if (options.type!=Modal.Type.Custom){
            options.callback?.let { cb ->
                findViewById<TextView>(R.id.ok).apply {
                    text = cb.displayText
                    setCompoundDrawablesWithIntrinsicBounds(cb.iconResourceId,0,0,0)
                    setOnClickListener {
                        if (cb.clickListener(this))
                            hide(bg,true)

                    }
                }
            }
        }
    }

    private fun setList(){
        if (options.type==Modal.Type.List){
            val doneBtnView=findViewById<TextView>(R.id.ok)
            findViewById<RecyclerView>(R.id.listScrollView).apply {
                if (adapter==null){
                    layoutManager=LinearLayoutManager(context)
                    adapter=ListAdapter(doneBtnView.typeface,options.list){ position->
                        hide(bg,true)
                    }
                }else
                    (adapter as ListAdapter).setMedicines(options.list)
            }
        }
    }

    private fun setProgress(){
        if (options.type==Modal.Type.Progress){
            findViewById<TextView>(R.id.percentage).text=options.progress.toString()
            val progress=findViewById<ProgressBar>(R.id.progress)
            progress.isIndeterminate=options.progress==0

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                progress.setProgress(options.progress,true)

            }else
                progress.progress=options.progress
        }
    }

    private fun setViewDirection(){
        background =
                when {
                    options.backgroundDrawable != null -> options.backgroundDrawable
                    options.direction == Modal.Direction.Bottom -> ResourcesCompat.getDrawable(resources,R.drawable.top_curved_header,null)
                    else -> ResourcesCompat.getDrawable(resources,R.drawable.bottom_curved_header,null)
                }
    }


    private fun blurEffect(set:Boolean){
        if (options.blurEnabled){
            //Do some blur magic..
            if(set){
                drawBlurBg()
                animateInBlurView()
            }

            else
                animateOutBlurView()
        }
    }

    private fun animateInBlurView() {
        blurBg.startAnimation(
                AnimationUtils.loadAnimation(context, R.anim.alpha_enter)
                        .apply {
                            duration = options.animationDuration
                        }
        )
    }

    private fun animateOutBlurView() {
        blurBg.startAnimation(
                AnimationUtils.loadAnimation(context, R.anim.alpha_exit)
                        .apply {
                            duration = options.animationDuration
                        }
        )
    }

    fun isShowing():Boolean = bg.parent != null

    override fun onChanged(obj: ModalObj) {
        val patches: MutableList<() -> Unit> = mutableListOf()

        if(
                obj.title!=options.title ||
                obj.message!=options.message ||
                obj.icon!=options.icon
        ){
            patches += {
                setHeader(bg)
            }
        }

        if(
                obj.callback!=options.callback ||
                obj.onDismiss!=options.onDismiss ||
                obj.onShow!=options.onShow
        ){
            patches += {
                setCallback()
            }
        }

        if(
                obj.list!=options.list
        ){
            patches += {
                setList()
            }
        }

        if(
                obj.progress!=options.progress
        ){
            patches += {
                setProgress()
            }
        }

        options = obj
        patches.forEach {
            it.invoke()
        }
    }

    fun update(patch: ModalObj.()->Unit) = with(options.copy()){
        patch()
        options.update.postValue(this)
    }
}